presto-main/src/main/java/com/facebook/presto/type/IpAddressOperators.java

```diff
@@ -0,0 +1,145 @@
+    {
+        return left.compareTo(right) < 0;
+    }
+
+    @ScalarOperator(LESS_THAN_OR_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean lessThanOrEqual(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) <= 0;
+    }
+
+    @ScalarOperator(GREATER_THAN)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean greaterThan(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) > 0;
+    }
+
+    @ScalarOperator(GREATER_THAN_OR_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean greaterThanOrEqual(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) >= 0;
+    }
+
+    @ScalarOperator(BETWEEN)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean between(@SqlType(StandardTypes.IPADDRESS) Slice value, @SqlType(StandardTypes.IPADDRESS) Slice min, @SqlType(StandardTypes.IPADDRESS) Slice max)
+    {
+        return min.compareTo(value) <= 0 && value.compareTo(max) <= 0;
+    }
+
+    @ScalarOperator(HASH_CODE)
+    @SqlType(StandardTypes.BIGINT)
+    public static long hashCode(@SqlType(StandardTypes.IPADDRESS) Slice value)
+    {
+        return XxHash64.hash(value);
+    }
+
+    @LiteralParameters("x")
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.IPADDRESS)
+    public static Slice castFromVarchartoIpAddress(@SqlType("varchar(x)") Slice slice)
+    {
+        try {
+            byte[] address = forString(slice.toStringUtf8()).getAddress();
+            byte[] bytes = new byte[16];
+            if (address.length == 4) {
+                bytes[10] = (byte) 0xff;
+                bytes[11] = (byte) 0xff;
+                arraycopy(address, 0, bytes, 12, 4);
+            }
+            else if (address.length == IP_ADDRESS_SIZE) {
+                arraycopy(address, 0, bytes, 0, IP_ADDRESS_SIZE);
+            }
+            else {
+                throw new PrestoException(INVALID_CAST_ARGUMENT, format("Can not cast '%s' to IPADDRESS", slice.toStringUtf8()));
+            }
+            return wrappedBuffer(bytes);
+        }
+        catch (IllegalArgumentException e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Can not cast '%s' to IPADDRESS", slice.toStringUtf8()));
+        }
+    }
+
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice castFromIpAddressToVarchar(@SqlType(StandardTypes.IPADDRESS) Slice slice)
+    {
+        try {
+            return utf8Slice(toAddrString(InetAddress.getByAddress(slice.getBytes())));
+        }
+        catch (UnknownHostException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Failed to load IP address '%s'", slice.toStringUtf8()));
```

Comment:
Don't call `toStringUtf8()` on arbitrary binary data, as it will crash if it's not UTF-8. Also, I'm not sure `GENERIC_INTERNAL_ERROR` is best since this bad data could come from a connector. How about
```java
throw new PrestoException(INVALID_CAST_ARGUMENT, "Invalid IP address binary length: " + slice.length(), e);